src/python/pants/backend/python/lint/black/rules.py

```diff
@@ -158,7 +158,7 @@ def generate_args(
     return FmtResult.from_execute_process_result(result)
 
 
-@rule(name="Lint using Black")
+@named_rule(name="black_lint", desc="Lint using Black", rule_type=NamedRuleType("lint"))
```

Comment:
This makes sense I think; it's not hard to rename a python function.

Reply:
Great point, Stu. I agree.

Reply:
@stuhood in the process of implementing this today, we realized that requiring that the canonical name of a rule always be the function name will make it difficult to refactor rules in the future once there's code that depends on a specific canonical name string. I updated this PR to keep an overrideable name, but change the kwarg from `name` to `canonical_name` to try to indicate that it should be used sparingly; and also renamed all the `fmt` and `list` functions among the python linters/formatters to use more descriptive function names and obviate the need to give those rules a `canonical_name` annotation.

Reply:
In that case, the names of the rules should probably not be a public API that consumers depend on.

Part of the reason the engine is so type-centric is that "types" are expected to be the public API that is exposed for plugins to consume.

Reply:
> Part of the reason the engine is so type-centric is that "types" are expected to be the public API that is exposed for plugins to consume, and types can be individually deprecated (rather than functions).

In the context of work units, I disagree with using types to generate the unique identifier for a rule. That's a leak of the implementation that your every day user looking at reports on their run does not care about or know about. For example, does an end user really care that the rule `isort/rules.py:fmt` took `IsortSetup` as a parameter?

It's also very common for us to add new types or even rename types. That seems brittle to use the name of every type to generate the stable name for a workunit.

Reply:
> It's also very common for us to add new types or to rename types. That seems brittle to use the name of every type to generate the stable name for a workunit.

Once a deprecation policy is in place for plugins, this will need to become less common, and the shape of unions will need to stabilize. That doesn't apply to the internal implementation details of the plugin, but you could imagine that it would apply at boundaries like this one.

> In the context of work units, I disagree with using types to generate the unique identifier for a rule.

If the goal of the string name is to uniquely identify a rule to an end user, it doesn't sound like it would be possible to actually refactor things in any meaningful way. If you're saying that you might want to refactor rules such that the name `black_fmt` was attached to a rule that did something else... what might that rule actually be called?

I can understand exposing a simple slug. I'm just not sure I can see the case in which you'd want to "name" a rule something other than what it does. If it's de-emphasized in favor of defaulting to using the rule name, then I guess that's fine.

Reply:
@stuhood I think a simple slug is basically what we want clients of workunits to have available, and it's fine for that to be the rule function name in most cases (we were also thinking about using the fully-qualified module + function name, to avoid having to make rule authors care about what unrelated rule functions to the code they're working on might be called, but that makes that slug longer and also makes it harder to refactor rules, so we dropped that idea). The reason we want to keep the ability to override the slug name is because we don't want to constrain how rule authors can architect their functions, once people have written code against those slug names as exposed in workunit JSON, including in ways I can't think of at the moment. Deemphasizing the `canonical_name` option and defaulting to the rule name makes sense to me and it's what the code currently does.

Reply:
+1 to needing the override being a special case that 95% of rule authors can ignore. I like that `canonical_name` is much longer and more precise than `name` for this reason.

Reply:
Fine with me.